stdlib/public/SDK/Foundation/Data.swift

```diff
@@ -844,221 +528,1229 @@ internal final class _DataStorage {
+            return storage.bridgedReference(self.range)
+        }
+        
+        @inlinable
+        mutating func resetBytes(in range: Range<Int>) {
+            precondition(range.lowerBound <= endIndex, "index \(range.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            ensureUniqueReference()
+            storage.resetBytes(in: range)
+            if slice.range.upperBound < range.upperBound {
+                slice.range = slice.range.lowerBound..<range.upperBound
+            }
+        }
+        
+        @inlinable
+        mutating func replaceSubrange(_ subrange: Range<Index>, with bytes: UnsafeRawPointer, count cnt: Int) {
+            precondition(startIndex <= subrange.lowerBound, "index \(subrange.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(subrange.lowerBound <= endIndex, "index \(subrange.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(startIndex <= subrange.upperBound, "index \(subrange.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(subrange.upperBound <= endIndex, "index \(subrange.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            
+            let nsRange = NSRange(location: subrange.lowerBound, length: subrange.upperBound - subrange.lowerBound)
+            ensureUniqueReference()
+            let upper = range.upperBound
+            storage.replaceBytes(in: nsRange, with: bytes, length: cnt)
+            let resultingUpper = upper - nsRange.length + cnt
+            slice.range = slice.range.lowerBound..<resultingUpper
+        }
+        
+        @inlinable
+        func copyBytes(to pointer: UnsafeMutableRawPointer, from range: Range<Int>) {
+            precondition(startIndex <= range.lowerBound, "index \(range.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(range.lowerBound <= endIndex, "index \(range.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(startIndex <= range.upperBound, "index \(range.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(range.upperBound <=     endIndex, "index \(range.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            
+            storage.withUnsafeBytes(in: range.lowerBound..<range.upperBound) {
+                UnsafeMutableRawBufferPointer(start: pointer, count: range.upperBound - range.lowerBound).copyMemory(from: $0)
+            }
+        }
+        
+        @inlinable
+        var hashValue: Int {
+            let hashRange = startIndex..<Swift.min(startIndex + 80, endIndex)
+            return storage.withUnsafeBytes(in: hashRange) {
+                return Int(bitPattern: CFHashBytes(UnsafeMutablePointer(mutating: $0.baseAddress?.assumingMemoryBound(to: UInt8.self)), CFIndex($0.count)))
+            }
+        }
+    }
+    
+    @usableFromInline
```

Comment:
You should add a @_frozen attribute to the enum. Otherwise it will not be fully optimized.

Reply:
Just managed to benchmark locally — looks like this made a **huge** difference on our results. We're seeing performance gains on all of the subscript benchmarks we were previously lagging on.

Reply:
Fantastic! Glad to hear our strategy is working. Are we seeing improvements across the board yet? Or are there some stragglers still?

Reply:
We do have stragglers — largest ones on my machine are `DataCopyBytesMedium`, `DataSetCountMedium`, and `DataSetCountLarge`, but those are the only ones past 40% regression (and are hovering at 100% regression, which is significantly less than what we were seeing). It should now be easier to hunt down the causes of these specifically and address them.

(I did also change a fair bit of the underlying protocols here, so if there's an extension I missed or something, that could contribute to the loss.)

Reply:
Ok, sounds good. We can dive into some profiles of those on Monday.

Reply:
To give you a very rough feeling for the changes on my machine:

```
N	46
min	-115.783
q1	-9.7447
median	-1.8503
q3	13.1967
max	98.2076
sum	233.461
mean	5.07525
stddev	48.4748
stderr	7.14721
variance	2349.8
```